# Credit Card Clustering

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
#from sklearn.metrics import 

### Step 1: Importing & Understanding the dataset

In [2]:
credit_card_df = pd.read_csv('datasets/Credit_card_dataset.csv')
credit_card_df.head()

CUST_ID  BALANCE_FREQUENCY  PURCHASES     PAYMENTS  CREDIT_LIMIT  \
0  C10001           0.818182      95.40   201.802084        1000.0   
1  C10002           0.909091       0.00  4103.032597        7000.0   
2  C10003           1.000000     773.17   622.066742        7500.0   
3  C10004           0.636364    1499.00     0.000000        7500.0   
4  C10005           1.000000      16.00   678.334763        1200.0   

   CASH_ADVANCE  
0      0.000000  
1   6442.945483  
2      0.000000  
3    205.788017  
4      0.000000

In [3]:
credit_card_df.shape

(8950, 6)

In [4]:
credit_card_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8950 entries, 0 to 8949
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CUST_ID            8950 non-null   object 
 1   BALANCE_FREQUENCY  8950 non-null   float64
 2   PURCHASES          8950 non-null   float64
 3   PAYMENTS           8950 non-null   float64
 4   CREDIT_LIMIT       8949 non-null   float64
 5   CASH_ADVANCE       8950 non-null   float64
dtypes: float64(5), object(1)
memory usage: 419.7+ KB


In [6]:
credit_card_df.describe(include='all')

CUST_ID  BALANCE_FREQUENCY     PURCHASES      PAYMENTS  CREDIT_LIMIT  \
count     8950        8950.000000   8950.000000   8950.000000   8949.000000   
unique    8950                NaN           NaN           NaN           NaN   
top     C10001                NaN           NaN           NaN           NaN   
freq         1                NaN           NaN           NaN           NaN   
mean       NaN           0.877271   1003.204834   1733.143852   4494.449450   
std        NaN           0.236904   2136.634782   2895.063757   3638.815725   
min        NaN           0.000000      0.000000      0.000000     50.000000   
25%        NaN           0.888889     39.635000    383.276166   1600.000000   
50%        NaN           1.000000    361.280000    856.901546   3000.000000   
75%        NaN           1.000000   1110.130000   1901.134317   6500.000000   
max        NaN           1.000000  49039.570000  50721.483360  30000.000000   

        CASH_ADVANCE  
count    8950.000000  
unique           NaN  
top              NaN  
freq             NaN  
mean      978.871112  
std      2097.163877  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%      1113.821139  
max     47137.211760

In [7]:
# check if there are blank values
credit_card_df.isnull().sum()

CUST_ID              0
BALANCE_FREQUENCY    0
PURCHASES            0
PAYMENTS             0
CREDIT_LIMIT         1
CASH_ADVANCE         0
dtype: int64

In [9]:
# check duplicated values
credit_card_df.duplicated().sum()

np.int64(0)

There are 8950 rows with 6 columns. The columns are mostly numerical with only the CUST_ID is categorical.
There is 1 missing value in the credit_limit column
There are no duplicates in the dataframe

### Step 2: Data Preprocessing

#### 2.1 Filling in missing values on CREDIT_LIMIT

In [14]:
# Filling in missing values on CREDIT_LIMIT
df = credit_card_df.copy()
df.columns = df.columns.str.lower()
missing = df[df.credit_limit.isnull()]
missing

cust_id  balance_frequency  purchases  payments  credit_limit  \
5203  C15349           0.166667        0.0  9.040017           NaN   

      cash_advance  
5203    186.853063

In [18]:
df[df.payments < 1].value_counts()

cust_id  balance_frequency  purchases  payments  credit_limit  cash_advance
C10004   0.636364           1499.00    0.0       7500.0        205.788017      1
C16963   0.545455           21.99      0.0       4000.0        0.000000        1
C15924   0.000000           1021.56    0.0       3000.0        0.000000        1
C15944   0.090909           0.00       0.0       1800.0        93.468447       1
C15949   0.000000           102.72     0.0       1500.0        0.000000        1
                                                                              ..
C11379   0.000000           300.00     0.0       2500.0        0.000000        1
C11624   1.000000           0.00       0.0       3000.0        1820.800715     1
C11675   0.090909           0.00       0.0       12000.0       1125.868686     1
C11705   0.090909           0.00       0.0       1000.0        575.193978      1
C19185   0.833333           1012.73    0.0       4000.0        0.000000        1
Name: count, Length: 242, dtype: 

In [25]:
df[df.payments < 1].credit_limit.mean()

np.float64(3908.471074380165)

Given there are many factors that affect credit limits, we'll assign the missing values a credit limit based off the average of the limits for people with payments less than 1. The reason we're utilising payments is because we're able to get lesser groupings as opposed to the other columns which are quite random

In [27]:
df.loc[5203, 'credit_limit'] = 3908.47

In [29]:
df.credit_limit.isnull().sum()

np.int64(0)

#### 2.2 Data analysis

1. Does balance frequency influence the credit limit amount?
2. Does making high purchase amounts and payments frequency affect the cash advance and credit limit?

#### 2.3 Checking and handling outliers

#### 2.4 Encoding the categorical columns

#### 2.5 Checking the correlation in our data

#### 2.6 If need be, we'll standardize/normalize our data

### Step 3: Training our models